src/vm/class.cpp

```diff
@@ -3198,6 +3198,251 @@ void EEClass::SetPackableField(EEClassFieldId eField, DWORD dwValue)
+
+            // pSortWalk now points to the target location for new LayoutRawFieldInfo*.
+            MoveMemory(pSortWalk + 1, pSortWalk, (pSortArrayEnd - pSortWalk) * sizeof(LayoutRawFieldInfo*));
+            *pSortWalk = &pFieldInfoArray[i];
+            pSortArrayEnd++;
+        }
+    }
+    else // no sorting for explicit layout
+    {
+        for (ULONG i = 0; i < cInstanceFields; i++)
+        {
+            if (pFieldInfoArray[i].m_MD != mdFieldDefNil)
+            {
+                if (pFieldInfoArray[i].m_nativePlacement.m_offset == (UINT32)-1)
+                {
+                    LPCUTF8 szFieldName;
+                    if (FAILED(pInternalImport->GetNameOfFieldDef(pFieldInfoArray[i].m_MD, &szFieldName)))
+                    {
+                        szFieldName = "Invalid FieldDef record";
+                    }
+                    pModule->GetAssembly()->ThrowTypeLoadException(pInternalImport,
+                        cl,
+                        szFieldName,
+                        IDS_CLASSLOAD_NSTRUCT_EXPLICIT_OFFSET);
+                }
+                else if ((INT)pFieldInfoArray[i].m_nativePlacement.m_offset < 0)
+                {
+                    LPCUTF8 szFieldName;
+                    if (FAILED(pInternalImport->GetNameOfFieldDef(pFieldInfoArray[i].m_MD, &szFieldName)))
+                    {
+                        szFieldName = "Invalid FieldDef record";
+                    }
+                    pModule->GetAssembly()->ThrowTypeLoadException(pInternalImport,
+                        cl,
+                        szFieldName,
+                        IDS_CLASSLOAD_NSTRUCT_NEGATIVE_OFFSET);
+                }
+            }
+
+            *pSortArrayEnd = &pFieldInfoArray[i];
+            pSortArrayEnd++;
+        }
+    }
+}
+
+void EEClassLayoutInfo::CalculateSizeAndFieldOffsets(
+    const UINT32 parentSize,
+    ULONG numInstanceFields,
+    BOOL fExplicitOffsets,
+    LayoutRawFieldInfo* const* pSortedFieldInfoArray,
+    ULONG classSizeInMetadata,
+    BYTE packingSize,
+    BYTE parentAlignmentRequirement,
+    BOOL calculatingNativeLayout,
+    EEClassLayoutInfo* pEEClassLayoutInfoOut
+)
+{
+    UINT32 cbCurOffset = parentSize;
+    BYTE LargestAlignmentRequirement = max(1, min(packingSize, parentAlignmentRequirement));
+
+    // Start with the size inherited from the parent (if any).
+    unsigned calcTotalSize = parentSize;
+
+    LayoutRawFieldInfo* const* pSortWalk;
+    ULONG i;
+    for (pSortWalk = pSortedFieldInfoArray, i = numInstanceFields; i; i--, pSortWalk++)
+    {
+        LayoutRawFieldInfo* pfwalk = *pSortWalk;
+        RawFieldPlacementInfo* placementInfo;
+
+        if (calculatingNativeLayout)
+        {
+            placementInfo = &pfwalk->m_nativePlacement;
+        }
+        else
+        {
+            placementInfo = &pfwalk->m_managedPlacement;
+        }
+
+        BYTE alignmentRequirement = placementInfo->m_alignment;
+
+        if (!(alignmentRequirement == 1 ||
```

Comment:
This seems like a good place for a `switch` instead of the awkward compound `||` expression.